diff --git a/compose/ui/ui-uikit/src/iosMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.h b/compose/ui/ui-uikit/src/iosMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.h
index 5387115422eae..646b6d22ae463 100644
--- a/compose/ui/ui-uikit/src/iosMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.h
+++ b/compose/ui/ui-uikit/src/iosMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.h
@@ -33,6 +33,15 @@ NS_ASSUME_NONNULL_BEGIN
 /// Indicates that trait interface style trait changed
 - (void)userInterfaceStyleDidChange CMP_ABSTRACT_FUNCTION;
 
+// Overrides of the UIViewController (UIViewControllerRotation) extension
+- (BOOL)shouldAutorotate;
+- (UIInterfaceOrientation)preferredInterfaceOrientationForPresentation;
+- (UIInterfaceOrientationMask)supportedInterfaceOrientations;
+
+// Overrides of the UIViewController (UIContainerViewControllerProtectedMethods) extension
+- (nullable UIViewController *)childViewControllerForStatusBarStyle;
+- (nullable UIViewController *)childViewControllerForStatusBarHidden;
+
 @end
 
 NS_ASSUME_NONNULL_END
diff --git a/compose/ui/ui-uikit/src/iosMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.m b/compose/ui/ui-uikit/src/iosMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.m
index c0679246b1a31..a97438372a6a8 100644
--- a/compose/ui/ui-uikit/src/iosMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.m
+++ b/compose/ui/ui-uikit/src/iosMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.m
@@ -182,6 +182,26 @@ - (void)userInterfaceStyleDidChange {
     CMP_ABSTRACT_FUNCTION_CALLED
 }
 
+- (BOOL)shouldAutorotate {
+    return [super shouldAutorotate];
+}
+
+- (UIInterfaceOrientation)preferredInterfaceOrientationForPresentation {
+    return [super preferredInterfaceOrientationForPresentation];
+}
+
+- (UIInterfaceOrientationMask)supportedInterfaceOrientations {
+    return [super supportedInterfaceOrientations];
+}
+
+- (UIViewController *)childViewControllerForStatusBarStyle {
+    return [super childViewControllerForStatusBarStyle];
+}
+
+- (UIViewController *)childViewControllerForStatusBarHidden {
+    return [super childViewControllerForStatusBarHidden];
+}
+
 - (void)dealloc {
     if (_lifecycleState == CMPComposeContainerLifecycleStateStarted) {
         [self viewControllerDidLeaveWindowHierarchy];
diff --git a/compose/ui/ui/src/iosMain/kotlin/androidx/compose/ui/scene/ComposeLayersViewController.ios.kt b/compose/ui/ui/src/iosMain/kotlin/androidx/compose/ui/scene/ComposeLayersViewController.ios.kt
index 37a53a14a775a..d0c311f88bbc5 100644
--- a/compose/ui/ui/src/iosMain/kotlin/androidx/compose/ui/scene/ComposeLayersViewController.ios.kt
+++ b/compose/ui/ui/src/iosMain/kotlin/androidx/compose/ui/scene/ComposeLayersViewController.ios.kt
@@ -21,6 +21,8 @@ import androidx.compose.ui.graphics.asComposeCanvas
 import androidx.compose.ui.platform.PlatformWindowContext
 import androidx.compose.ui.uikit.addLayoutConstraintsToMatch
 import androidx.compose.ui.uikit.embedSubview
+import androidx.compose.ui.uikit.utils.CMPComposeContainerLifecycleDelegateProtocol
+import androidx.compose.ui.uikit.utils.CMPViewController
 import androidx.compose.ui.util.fastForEach
 import androidx.compose.ui.viewinterop.UIKitInteropTransaction
 import androidx.compose.ui.window.ComposeContainerView
@@ -41,6 +43,8 @@ import org.jetbrains.skia.Canvas
 import platform.CoreGraphics.CGPoint
 import platform.CoreGraphics.CGSize
 import platform.UIKit.UIEvent
+import platform.UIKit.UIInterfaceOrientation
+import platform.UIKit.UIInterfaceOrientationMask
 import platform.UIKit.UIScreen
 import platform.UIKit.UIView
 import platform.UIKit.UIViewController
@@ -48,6 +52,12 @@ import platform.UIKit.UIViewControllerTransitionCoordinatorProtocol
 import platform.UIKit.UIWindow
 import platform.UIKit.UIWindowLevelAlert
 import platform.UIKit.UIWindowLevelNormal
+import platform.UIKit.childViewControllerForStatusBarHidden
+import platform.UIKit.childViewControllerForStatusBarStyle
+import platform.UIKit.preferredInterfaceOrientationForPresentation
+import platform.UIKit.shouldAutorotate
+import platform.UIKit.supportedInterfaceOrientations
+import platform.darwin.NSObject
 
 /**
  * A class responsible for managing and rendering [UIKitComposeSceneLayer]s.
@@ -55,7 +65,7 @@ import platform.UIKit.UIWindowLevelNormal
 internal class ComposeLayersViewController(
     useSeparateRenderThreadWhenPossible: Boolean,
     private val coroutineContext: CoroutineContext
-): UIViewController(nibName = null, bundle = null) {
+): CMPViewController(lifecycleDelegate = EmptyComposeContainerLifecycleDelegate()) {
     val windowContext = PlatformWindowContext()
 
     private val window = LayersWindow()
@@ -166,7 +176,7 @@ internal class ComposeLayersViewController(
         val transaction = layer.retrieveInteropTransaction()
 
         if (this.layers.isEmpty()) {
-            // It was the last layer, remove the view and executed the actions immediately
+            // It was the last layer, remove the view and execute the actions immediately
             hide()
 
             transaction.actions.fastForEach { it.invoke() }
@@ -214,7 +224,7 @@ internal class ComposeLayersViewController(
 
     /**
      * Iterate through existing layers and merge their interop transactions to be consumed by the
-     * [MetalView], also include transactions of the layers that were removed and are not
+     * [MetalView], also including transactions of the layers that were removed and are not
      * present in [layers] anymore.
      */
     private fun retrieveAndMergeInteropTransactions(): UIKitInteropTransaction {
@@ -263,6 +273,38 @@ internal class ComposeLayersViewController(
         }
     }
 
+    override fun viewControllerDidEnterWindowHierarchy() {}
+
+    override fun viewControllerDidLeaveWindowHierarchy() {}
+
+    override fun userInterfaceStyleDidChange() {}
+
+    override fun preferredInterfaceOrientationForPresentation(): UIInterfaceOrientation {
+        return referenceWindow?.rootViewController?.preferredInterfaceOrientationForPresentation()
+            ?: super.preferredInterfaceOrientationForPresentation()
+    }
+
+    override fun supportedInterfaceOrientations(): UIInterfaceOrientationMask {
+        return referenceWindow?.rootViewController?.supportedInterfaceOrientations()
+            ?: super.supportedInterfaceOrientations()
+    }
+
+    override fun shouldAutorotate(): Boolean {
+        return referenceWindow?.rootViewController?.shouldAutorotate() ?: super.shouldAutorotate()
+    }
+
+    override fun childViewControllerForStatusBarStyle(): UIViewController? {
+        return referenceWindow?.rootViewController?.childViewControllerForStatusBarStyle()
+            ?: referenceWindow?.rootViewController
+            ?: super.childViewControllerForStatusBarStyle()
+    }
+
+    override fun childViewControllerForStatusBarHidden(): UIViewController? {
+        return referenceWindow?.rootViewController?.childViewControllerForStatusBarHidden()
+            ?: referenceWindow?.rootViewController
+            ?: super.childViewControllerForStatusBarHidden()
+    }
+
     /**
      * Animates the layout transition of layers.
      * See [androidx.compose.ui.scene.ComposeHostingViewController.animateSizeTransition]
@@ -327,9 +369,9 @@ internal class ComposeLayersViewController(
     }
 }
 
-private class LayersWindow: UIWindow(frame = UIScreen.mainScreen.bounds) {
+internal class LayersWindow: UIWindow(frame = UIScreen.mainScreen.bounds) {
     override fun hitTest(point: CValue<CGPoint>, withEvent: UIEvent?): UIView? {
-        // Hit-testing only the Compose view or any view that located on top of it.
+        // Hit-testing only the Compose view or any view that is located on top of it.
         for (subview in subviews.reversed()) {
             subview as UIView
             val isDescendantOfComposeView = rootViewController?.view?.isDescendantOfView(subview)
@@ -344,3 +386,10 @@ private class LayersWindow: UIWindow(frame = UIScreen.mainScreen.bounds) {
         return null
     }
 }
+
+private class EmptyComposeContainerLifecycleDelegate: NSObject(),
+    CMPComposeContainerLifecycleDelegateProtocol {
+    override fun composeContainerWillDealloc() {}
+    override fun composeContainerWillAppear() {}
+    override fun composeContainerDidDisappear() {}
+}
diff --git a/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/integrations/ComposeHostingViewIntegrationTest.kt b/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/integrations/ComposeHostingViewIntegrationTest.kt
index cf3bdba83b2ca..ee5191e57a9e6 100644
--- a/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/integrations/ComposeHostingViewIntegrationTest.kt
+++ b/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/integrations/ComposeHostingViewIntegrationTest.kt
@@ -60,6 +60,8 @@ class ComposeHostingViewIntegrationTest {
         UIKitInstrumentedTest.waitUntil("Wait until compose view being disposed") {
             composeView.rootRedrawer == null
         }
+
+        appDelegate.cleanUp()
     }
 
     @Test
@@ -94,5 +96,7 @@ class ComposeHostingViewIntegrationTest {
 
         assertNotNull(composeView.rootRedrawer, "ComposeView should be alive")
         assertEquals(1, compositionsCount, "Compose view should not have extra recompositions")
+
+        appDelegate.cleanUp()
     }
 }
\ No newline at end of file
diff --git a/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/integrations/ComposeLaunchTest.kt b/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/integrations/ComposeLaunchTest.kt
index d3ddf92c005de..3dec1d19ee90c 100644
--- a/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/integrations/ComposeLaunchTest.kt
+++ b/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/integrations/ComposeLaunchTest.kt
@@ -89,6 +89,8 @@ class ComposeLaunchTest {
         image.forEachPixel { _, _, color ->
             assertEquals(Color.Blue, color, "Expected to draw blue background")
         }
+
+        appDelegate.cleanUp()
     }
 
     @OptIn(ExperimentalForeignApi::class)
@@ -124,5 +126,7 @@ class ComposeLaunchTest {
         image.forEachPixel { _, _, color ->
             assertEquals(Color.Blue, color, "Expected to draw blue background")
         }
+
+        appDelegate.cleanUp()
     }
 }
diff --git a/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/scene/ComposeLayersViewControllerTest.kt b/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/scene/ComposeLayersViewControllerTest.kt
new file mode 100644
index 0000000000000..b1da115f74e01
--- /dev/null
+++ b/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/scene/ComposeLayersViewControllerTest.kt
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2026 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.scene
+
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.size
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.snapshots.Snapshot
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.test.MockAppDelegate
+import androidx.compose.ui.test.UIKitInstrumentedTest
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.uikit.utils.CMPComposeContainerLifecycleDelegateProtocol
+import androidx.compose.ui.uikit.utils.CMPViewController
+import androidx.compose.ui.window.ComposeUIViewController
+import androidx.compose.ui.window.Dialog
+import androidx.compose.ui.window.Popup
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+import platform.UIKit.UIInterfaceOrientation
+import platform.UIKit.UIInterfaceOrientationLandscapeRight
+import platform.UIKit.UIInterfaceOrientationMask
+import platform.UIKit.UIInterfaceOrientationMaskAll
+import platform.UIKit.UIInterfaceOrientationMaskPortraitUpsideDown
+import platform.UIKit.UIInterfaceOrientationPortrait
+import platform.UIKit.UIStatusBarStyle
+import platform.UIKit.UIStatusBarStyleDarkContent
+import platform.UIKit.UIStatusBarStyleDefault
+import platform.UIKit.addChildViewController
+import platform.UIKit.childViewControllerForStatusBarHidden
+import platform.UIKit.childViewControllerForStatusBarStyle
+import platform.UIKit.didMoveToParentViewController
+import platform.UIKit.preferredInterfaceOrientationForPresentation
+import platform.UIKit.shouldAutorotate
+import platform.UIKit.supportedInterfaceOrientations
+import platform.darwin.NSObject
+
+class ComposeLayersViewControllerTest {
+    @Test
+    fun testOrientationAndStatusBarValues() {
+        val appDelegate = MockAppDelegate()
+        val viewController = TestComposeContainerViewController { Dialog(onDismissRequest = {}) {} }
+        appDelegate.setUpWindow(viewController = viewController)
+        viewController.waitForIdle()
+
+        viewController.overrideShouldAutorotate = true
+        viewController.overridePreferredInterfaceOrientationForPresentation =
+            UIInterfaceOrientationPortrait
+        viewController.overrideSupportedInterfaceOrientations = UIInterfaceOrientationMaskAll
+
+        viewController.overridePreferredStatusBarStyle = UIStatusBarStyleDefault
+        viewController.overridePrefersStatusBarHidden = false
+
+        val layersWindow = appDelegate.findLayersWindow()
+
+        assertEquals(true, layersWindow.rootViewController?.shouldAutorotate)
+        assertEquals(
+            UIInterfaceOrientationPortrait,
+            layersWindow.rootViewController?.preferredInterfaceOrientationForPresentation
+        )
+        assertEquals(
+            UIInterfaceOrientationMaskAll,
+            layersWindow.rootViewController?.supportedInterfaceOrientations
+        )
+        assertEquals(
+            UIStatusBarStyleDefault,
+            layersWindow.rootViewController?.childViewControllerForStatusBarStyle?.preferredStatusBarStyle
+        )
+        assertEquals(
+            false,
+            layersWindow.rootViewController?.childViewControllerForStatusBarHidden?.prefersStatusBarHidden
+        )
+
+        viewController.overrideShouldAutorotate = false
+        viewController.overridePreferredInterfaceOrientationForPresentation =
+            UIInterfaceOrientationLandscapeRight
+        viewController.overrideSupportedInterfaceOrientations =
+            UIInterfaceOrientationMaskPortraitUpsideDown
+        viewController.overridePreferredStatusBarStyle = UIStatusBarStyleDarkContent
+        viewController.overridePrefersStatusBarHidden = true
+
+        assertEquals(false, layersWindow.rootViewController?.shouldAutorotate)
+        assertEquals(
+            UIInterfaceOrientationLandscapeRight,
+            layersWindow.rootViewController?.preferredInterfaceOrientationForPresentation
+        )
+        assertEquals(
+            UIInterfaceOrientationMaskPortraitUpsideDown,
+            layersWindow.rootViewController?.supportedInterfaceOrientations
+        )
+
+        assertEquals(
+            UIStatusBarStyleDarkContent,
+            layersWindow.rootViewController?.childViewControllerForStatusBarStyle?.preferredStatusBarStyle
+        )
+        assertEquals(
+            true,
+            layersWindow.rootViewController?.childViewControllerForStatusBarHidden?.prefersStatusBarHidden
+        )
+
+        appDelegate.cleanUp()
+    }
+
+    @Test
+    fun testPopupOrientationAndStatusBarValues() {
+        val appDelegate = MockAppDelegate()
+        val viewController = TestComposeContainerViewController {
+            Popup {
+                Box(Modifier.size(10.dp))
+            }
+        }
+        appDelegate.setUpWindow(viewController = viewController)
+        viewController.waitForIdle()
+
+        viewController.overrideShouldAutorotate = true
+        viewController.overridePreferredInterfaceOrientationForPresentation =
+            UIInterfaceOrientationPortrait
+        viewController.overrideSupportedInterfaceOrientations = UIInterfaceOrientationMaskAll
+
+        viewController.overridePreferredStatusBarStyle = UIStatusBarStyleDefault
+        viewController.overridePrefersStatusBarHidden = false
+
+        val layersWindow = appDelegate.findLayersWindow()
+
+        assertEquals(
+            true,
+            layersWindow.rootViewController?.shouldAutorotate,
+            "Autorotation must be enabled"
+        )
+        assertEquals(
+            UIInterfaceOrientationPortrait,
+            layersWindow.rootViewController?.preferredInterfaceOrientationForPresentation
+        )
+        assertEquals(
+            UIInterfaceOrientationMaskAll,
+            layersWindow.rootViewController?.supportedInterfaceOrientations
+        )
+        assertEquals(
+            UIStatusBarStyleDefault,
+            layersWindow.rootViewController?.childViewControllerForStatusBarStyle?.preferredStatusBarStyle
+        )
+        assertEquals(
+            false,
+            layersWindow.rootViewController?.childViewControllerForStatusBarHidden?.prefersStatusBarHidden,
+            "Status bar should not be hidden"
+        )
+
+        viewController.overrideShouldAutorotate = false
+        viewController.overridePreferredInterfaceOrientationForPresentation =
+            UIInterfaceOrientationLandscapeRight
+        viewController.overrideSupportedInterfaceOrientations =
+            UIInterfaceOrientationMaskPortraitUpsideDown
+        viewController.overridePreferredStatusBarStyle = UIStatusBarStyleDarkContent
+        viewController.overridePrefersStatusBarHidden = true
+
+        assertEquals(
+            false,
+            layersWindow.rootViewController?.shouldAutorotate,
+            "Autorotation must be disabled"
+        )
+        assertEquals(
+            UIInterfaceOrientationLandscapeRight,
+            layersWindow.rootViewController?.preferredInterfaceOrientationForPresentation
+        )
+        assertEquals(
+            UIInterfaceOrientationMaskPortraitUpsideDown,
+            layersWindow.rootViewController?.supportedInterfaceOrientations
+        )
+
+        assertEquals(
+            UIStatusBarStyleDarkContent,
+            layersWindow.rootViewController?.childViewControllerForStatusBarStyle?.preferredStatusBarStyle
+        )
+        assertEquals(
+            true,
+            layersWindow.rootViewController?.childViewControllerForStatusBarHidden?.prefersStatusBarHidden,
+            "Status bar should be hidden"
+        )
+
+        appDelegate.cleanUp()
+    }
+
+    private fun MockAppDelegate.findLayersWindow(): LayersWindow {
+        val window = this@findLayersWindow.window?.windowScene?.windows?.mapNotNull {
+            it as? LayersWindow
+        }?.single()
+
+        assertNotNull(window, "${LayersWindow::class} not found in scene")
+        return window
+    }
+}
+
+private class TestComposeContainerViewController(
+    private val content: @Composable () -> Unit
+): CMPViewController(lifecycleDelegate = TestComposeContainerLifecycleDelegate()) {
+    var overrideShouldAutorotate: Boolean = true
+
+    var overridePreferredInterfaceOrientationForPresentation: UIInterfaceOrientation =
+        UIInterfaceOrientationPortrait
+
+    var overrideSupportedInterfaceOrientations: UIInterfaceOrientationMask =
+        UIInterfaceOrientationMaskAll
+
+    var overridePreferredStatusBarStyle: UIStatusBarStyle =
+        UIStatusBarStyleDefault
+
+    var overridePrefersStatusBarHidden: Boolean =
+        false
+
+    override fun preferredInterfaceOrientationForPresentation(): UIInterfaceOrientation =
+        overridePreferredInterfaceOrientationForPresentation
+
+    override fun supportedInterfaceOrientations(): UIInterfaceOrientationMask =
+        overrideSupportedInterfaceOrientations
+
+    override fun shouldAutorotate(): Boolean = overrideShouldAutorotate
+
+    override fun preferredStatusBarStyle(): UIStatusBarStyle = overridePreferredStatusBarStyle
+
+    override fun prefersStatusBarHidden(): Boolean = overridePrefersStatusBarHidden
+
+    val composeViewController = ComposeUIViewController(
+        configure = { enforceStrictPlistSanityCheck = false }
+    ) { content() } as ComposeHostingViewController
+
+    fun waitForIdle() =
+        UIKitInstrumentedTest.waitUntil {
+            !composeViewController.hasInvalidations() && !Snapshot.current.hasPendingChanges()
+        }
+
+    override fun viewDidLoad() {
+        super.viewDidLoad()
+
+        addChildViewController(composeViewController)
+        view.addSubview(composeViewController.view)
+        composeViewController.didMoveToParentViewController(this)
+    }
+
+    override fun viewDidLayoutSubviews() {
+        super.viewDidLayoutSubviews()
+
+        composeViewController.view.setFrame(view.bounds)
+    }
+
+    override fun viewControllerDidEnterWindowHierarchy() {}
+
+    override fun viewControllerDidLeaveWindowHierarchy() {}
+
+    override fun userInterfaceStyleDidChange() {}
+}
+
+private class TestComposeContainerLifecycleDelegate: NSObject(),
+    CMPComposeContainerLifecycleDelegateProtocol {
+    override fun composeContainerWillDealloc() {}
+    override fun composeContainerWillAppear() {}
+    override fun composeContainerDidDisappear() {}
+}
diff --git a/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/test/UIKitInstrumentedTest.kt b/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/test/UIKitInstrumentedTest.kt
index 46ae4df67b261..1b21b73a7201c 100644
--- a/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/test/UIKitInstrumentedTest.kt
+++ b/compose/ui/ui/src/uikitInstrumentedTest/kotlin/androidx/compose/ui/test/UIKitInstrumentedTest.kt
@@ -440,16 +440,25 @@ internal class MockAppDelegate: NSObject(), UIApplicationDelegateProtocol {
     fun setUpWindow(viewController: UIViewController) {
         UIApplication.sharedApplication().setDelegate(this)
 
+        val scene = UIApplication.sharedApplication().connectedScenes.first() as? UIWindowScene
+                ?: error("No window scene found")
+        val allWindows = scene.windows
+
         _window = UIWindow(frame = UIScreen.mainScreen.bounds)
         _window?.backgroundColor = UIColor.systemBackgroundColor
-        _window?.windowScene = UIApplication.sharedApplication().connectedScenes.first() as? UIWindowScene
+        _window?.windowScene = scene
 
         _window?.rootViewController = viewController
         _window?.makeKeyAndVisible()
+
+        allWindows.forEach {
+            (it as UIWindow).setHidden(true)
+            it.windowScene = null
+        }
     }
 
     fun cleanUp() {
-        val scene = UIApplication.sharedApplication().connectedScenes.first() as? UIWindowScene
+        val scene = _window?.windowScene ?: UIApplication.sharedApplication().connectedScenes.first() as? UIWindowScene
         val allWindows = scene?.windows ?: emptyList<UIWindow>()
 
         val window = UIWindow(frame = UIScreen.mainScreen.bounds)
@@ -469,6 +478,7 @@ internal class MockAppDelegate: NSObject(), UIApplicationDelegateProtocol {
 
         allWindows.forEach {
             (it as UIWindow).setHidden(true)
+            it.windowScene = null
         }
     }
 
